spec.md

```diff
@@ -1040,6 +919,83 @@ The width of each primitive operation is detailed in [@sec:primitive-operations]
 
 The width of constant integer expressions is detailed in their respective sections.
 
+### Reset Inference
+
+The following example shows an uninferred reset that will get inferred to a synchronous reset.
+
+``` firrtl
+input a : UInt<1>
+wire reset : Reset
+connect reset, a
+```
+
+After reset inference, `reset`{.firrtl} is inferred to the synchronous `UInt<1>`{.firrtl} type:
+
+``` firrtl
+input a : UInt<1>
+wire reset : UInt<1>
+connect reset, a
+```
+
+The following example demonstrates usage of an asynchronous reset.
+
+``` firrtl
+input clock : Clock
+input reset : AsyncReset
+input x : UInt<8>
+regreset y : UInt<8>, clock, reset, UInt(123)
+; ...
+```
+
+Inference rules are as follows:
+
+1. An uninferred reset driven by and/or driving only asynchronous resets will be inferred as asynchronous reset.
+1. An uninferred reset driven by and/or driving both asynchronous and synchronous resets is an error.
+1. Otherwise, the reset is inferred as synchronous (i.e. the uninferred reset is only invalidated or is driven by or drives only synchronous resets).
+
+`Reset`{.firrtl}s, whether synchronous or asynchronous, can be cast to other types.
+Casting between reset types is also legal:
+
+``` firrtl
+input a : UInt<1>
+output y : AsyncReset
+output z : Reset
+wire r : Reset
+connect r, a
+connect y, asAsyncReset(r)
+connect z, asUInt(y)
+```
+
+See [@sec:primitive-operations] for more details on casting.
+
+### Probes and Type Inference
+
+Probe types do participate in global width and reset inference, but only in the direction of the reference itself (no inference in the other direction, even with force statements).
+Both inner types of the references used in a `define`{.firrtl} statement must be identical or the same type with the destination uninferred (this is checked recursively).
+Additionally, any contained reset type is similarly only inferred in the direction of the reference, even if it eventually reaches a known reset type.
+
+In the following example, the FIRRTL compiler will produce an error constrasted with inferring the input port as `AsyncReset`{.firrtl} if a direct connection was used:
```

Comment:
I would avoid discussions of a FIRRTL compiler and opt instead for language around legal and illegal FIRRTL circuits. Also, while there is de facto one ("the") FIRRTL compiler, there may be many.

Reply:
Rewrote this section after consulting @dtzSiFive.

It's still hand-wavy, but only due to the fact that all of type inference is one big hand-wave at this point. It should suffice for now.